### DISCUSSION TITLE
suspicious-fs (SuS-FS) Discussion thread | @Kartatz

### DISCUSSION THREAD

**Discussion by itsHanibee** [Dominium-Apum/kernel_xiaomi_chime#6] [Subsystems]
The previous discussion was taking place @ PRs #1 #2 #3, they are not appropriate for discussions. Visit these for the full history regarding this patch.

**Comment by itsHanibee** [comment]
The people most involved with this series of patches are @Kartatz @backslashxx @simonpunk @itsHanibee, you are encouraged to discuss patchsets and any changes moving forward right here.

**Reply by itsHanibee** [reply]
Important snippets or discussion;

@backslashxx;
> @itsHanibee @Kartatz
> 
> might be interesting [backslashxx/mojito_krenol@2a35a43](https://github.com/backslashxx/mojito_krenol/commit/2a35a43ecd806a145de132faaf93e05648633031)

==========================================

> > sweet, thanks for the heads up I'll look into it. I've noticed you added guarding too.
> 
> this might be easier though [tiann/KernelSU#1494](https://github.com/tiann/KernelSU/pull/1494)

==========================================

> thanks to simonpunk he extended susfs and now it is controllable with a module and now with cool shit added. this is real cool
> 
> SuSFS v2 / XT / whatever Heres full implementation, make sure to read as it also ships with a module to control it! https://gitlab.com/simonpunk/susfs4ksu Mirror: https://github.com/backslashxx/susfs4ksu
> 
> here is msm-4.14 sample implementation with fixes included Kernel side: https://github.com/backslashxx/mojito_krenol/commits/susfs-xt-impl/ KernelSU side: https://github.com/backslashxx/KernelSU/commits/11758-susfs/
> 
> Basic requirements are
> 
>     * full path_umount backport with definition on internal.h
> 
>     * KernelSU
> 
> 
> a few fixes and some new features. at first glance, heres whats new, kstat spoofing, extendable umounts and uname spoofing (like arter97 's kernel/sys.c change?)
> 
> thanks a fuckton for this @simonpunk
> 
> before: (w/ kartatz's initial version) ![image](https://private-user-images.githubusercontent.com/118538522/319414529-4905717c-9eb9-49c0-8fb8-75fa2739f0db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyMDE0ODksIm5iZiI6MTcxMjIwMTE4OSwicGF0aCI6Ii8xMTg1Mzg1MjIvMzE5NDE0NTI5LTQ5MDU3MTdjLTllYjktNDljMC04ZmI4LTc1ZmEyNzM5ZjBkYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwNFQwMzI2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03N2IwZjhmYjRhMzI3N2Y1N2VmNmY0NTMwNzVhMjgwNmJjZjgyMzBiMDk3ZDQ4ZjY0MWQ4ZTJmMzNlNDExZmYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.C0cPjhf3r7XS3HiekQNxUuaxWqrmupiHTpFVhO2JCI0) after: ![image](https://private-user-images.githubusercontent.com/118538522/319414562-a2700d1e-1968-4600-a187-d72d55c4a7e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIyMDE0ODksIm5iZiI6MTcxMjIwMTE4OSwicGF0aCI6Ii8xMTg1Mzg1MjIvMzE5NDE0NTYyLWEyNzAwZDFlLTE5NjgtNDYwMC1hMTg3LWQ3MmQ1NWM0YTdlMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwNFQwMzI2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMDgzMTNlNWMxODgzNGZjZGU2NzBkMDIwMDg3Njc4MWVhZmQ3MzI4ZWFlMGQ0ODZkOWQ3OThlODA3NzQ3Y2FhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NaxBYfh4CTD5BnNx-720PZO7RTl4drESZtUktIFFpwg)

**Comment by simonpunk** [comment]
@itsHanibee Thanks for having me here. Just a wrap up:
The code rely on KernelSU and some syscall / kernel functions differ on other kernel versions as well, it's still hard to address for other kernel versions. I don't have time to test it all, I think we should just let other interested dev to fork it and make their own patches?

Extra kernel features I have added since from original commit by @Kartatz:
1. better deal with error code returned from some syscall functions. (maybe still more to be done)
2. spoof getdents/getdents64 syscall for hiding the suspicious file/dir
3. kstat spoofing file/dir that are bind mounted or overlayed from maps and syscalls (like you don't want to umount it for target process and keep the changes, but still want to avoid being detected for the changed ino, dev, and change time.)
4. spoof uname syscall.

Thoughts on other improvements:
1. Use hashed list head instead of linked list?
2. How to better secure the prctl called by malicious process? maybe add a token check at arg4? like use a script to let user to change it in patches and ksu_susfs tool.
3. Are there still better ways to hide in maps? Author of shamiko once mentioned that he can hide the trace better in kernel side than in userspace module. Clearly the current spoof logic is not enough to hide better in maps, advises and suggestions will be much appreciated.
